src/Checkout/Helpers/ReserveStock.php

```diff
@@ -0,0 +1,176 @@
+				foreach ( $rows as $product_id => $quantity ) {
+					$this->reserve_stock_for_product( $product_id, $quantity, $order, $minutes );
+				}
+			}
+		} catch ( ReserveStockException $e ) {
+			$this->release_stock_for_order( $order );
+			throw $e;
+		}
+	}
+
+	/**
+	 * Release a temporary hold on stock for an order.
+	 *
+	 * @param \WC_Order $order Order object.
+	 */
+	public function release_stock_for_order( \WC_Order $order ) {
+		global $wpdb;
+
+		$wpdb->delete(
+			$wpdb->wc_reserved_stock,
+			array(
+				'order_id' => $order->get_id(),
+			)
+		);
+	}
+
+	/**
+	 * Reserve stock for a product by inserting rows into the DB.
+	 *
+	 * @throws ReserveStockException If a row cannot be inserted.
+	 *
+	 * @param int       $product_id Product ID which is having stock reserved.
+	 * @param int       $stock_quantity Stock amount to reserve.
+	 * @param \WC_Order $order Order object which contains the product.
+	 * @param int       $minutes How long to reserve stock in minutes.
+	 */
+	private function reserve_stock_for_product( $product_id, $stock_quantity, \WC_Order $order, $minutes ) {
+		global $wpdb;
+
+		$product_data_store       = \WC_Data_Store::load( 'product' );
+		$query_for_stock          = $product_data_store->get_query_for_stock( $product_id );
+		$query_for_reserved_stock = $this->get_query_for_reserved_stock( $product_id, $order->get_id() );
+
+		// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared, WordPress.DB.PreparedSQL.NotPrepared
+		$result = $wpdb->query(
+			$wpdb->prepare(
+				"
+				REPLACE INTO {$wpdb->wc_reserved_stock} ( order_id, product_id, stock_quantity, expires )
+				SELECT %d, %d, %d, ( NOW() + INTERVAL %d MINUTE ) from DUAL
+				WHERE ( $query_for_stock FOR UPDATE ) - ( $query_for_reserved_stock FOR UPDATE ) >= %d
+				",
+				$order->get_id(),
+				$product_id,
+				$stock_quantity,
+				$minutes,
+				$stock_quantity
+			)
+		);
+		// phpcs:enable WordPress.DB.PreparedSQL.InterpolatedNotPrepared, WordPress.DB.PreparedSQL.NotPrepared
+
+		if ( ! $result ) {
+			$product = wc_get_product( $product_id );
+			throw new ReserveStockException(
+				'product_not_enough_stock',
+				sprintf(
+					/* translators: %s: product name */
+					__( 'Not enough units of %s are available in stock to fulfil this order.', 'woocommerce' ),
+					$product ? $product->get_name() : '#' . $product_id
+				),
+				403
+			);
+		}
+	}
+
+	/**
+	 * Returns query statement for getting reserved stock of a product.
+	 *
+	 * @param int     $product_id Product ID.
+	 * @param integer $exclude_order_id Optional order to exclude from the results.
+	 * @return string|void Query statement.
+	 */
+	private function get_query_for_reserved_stock( $product_id, $exclude_order_id = 0 ) {
+		global $wpdb;
+		return $wpdb->prepare(
+			"
+			SELECT COALESCE( SUM( stock_table.`stock_quantity` ), 0 ) FROM $wpdb->wc_reserved_stock stock_table
+			LEFT JOIN $wpdb->posts posts ON stock_table.`order_id` = posts.ID
```

Comment:
this join is going to be expensive and will possibly lock lots of rows. it's probably be worth exploring if there is some reliable way to copy over statuses in this table itself.

Reply:
Are you saying this locks everything and not just `wc_reserved_stock`?

Reply:
@mikejolley it should also lock `posts` table, since it's getting read in the left join.

Reply:
Still compared with the old query this new one is better.

Reply:
@vedanshujain We do clear up locks after status changes so we *could* remove the join completely, but we should maybe leave it in place as an extra layer of protection, and only remove if justified.

Reply:
If we can remove the join I recommend we do that. Actually, I have been planning to test out this PR for about 2 weeks now, let me do it today. Sorry for the delay on this.,

Reply:
I'd prefer to err on the side of caution and leave it so only pending/draft orders hold stock, _unless_ we see issues come out of this in testing? Syncing, like you suggested earlier, is always going to be less reliable. I'm worried about stock being held and then failing to free up later on for some unknown reason.

Reply:
I did a load test, and the approach in this PR is way faster than what we have for coupons. I think we should have different tables for storing locks for coupons as well. I am always hitting deadlocks on coupon queries before I hit them on the stock query when I gradually keep on increasing the load, so I would say we are locking much fewer rows here.

> Syncing, like you suggested earlier, is always going to be less reliable.

Agree, let's keep as it is.